File: concepts/03_errors_and_resources/04_contracts_assertions_and_failure_signaling/nodejs/test_02_configuration_and_runtime_boundaries.mjs
// 断言配置与运行时契约边界。
// 共同问题：开发断言能否被配置移除；类型约束是否检查运行时值；
// 公共 API 应采用什么稳定失败通道。
//
// polyglot-family: errors_and_resources
// polyglot-concept: contracts_assertions_and_failure_signaling
// polyglot-related: languages/nodejs/language/test_004_control_flow_completion_records_and_errors.mjs

import assert from 'node:assert/strict';
import test from 'node:test';

function positive(value) {
  if (typeof value !== 'number') {
    throw new TypeError('must be a number');
  }
  if (!Number.isInteger(value) || value <= 0) {
    throw new RangeError('must be a positive integer');
  }
  return value;
}

test('node:assert 是运行时函数而非可删除的语言语句', () => {
  assert.throws(
    () => assert.ok(false, 'broken invariant'),
    {
      name: 'AssertionError',
      message: /broken invariant/,
    },
  );

  // Node.js 没有与 Python -O 或 C++ NDEBUG 对应的语言级删除规则；打包器转换不属于
  // ECMAScript/Node 运行时保证。公共输入错误仍不应伪装成 AssertionError。
});

test('动态函数签名不替代运行时类型和值校验', () => {
  assert.equal(positive(2), 2);
  assert.throws(() => positive('2'), TypeError);
  assert.throws(() => positive(-2), RangeError);

  // TypeScript 可在外部构建阶段提供静态约束，但发给 Node.js 的 JavaScript 仍须保护
  // 不可信边界；正值约束也不是普通类型签名能够表达的运行时事实。
});

File: concepts/03_errors_and_resources/04_contracts_assertions_and_failure_signaling/python/test_02_configuration_and_runtime_boundaries.py
"""断言配置与运行时契约边界。

共同问题：开发断言能否被配置移除；类型约束是否检查运行时值；
公共 API 应采用什么稳定失败通道。
"""

# polyglot-family: errors_and_resources
# polyglot-concept: contracts_assertions_and_failure_signaling
# polyglot-related: languages/python/language/test_009_exception_handling_and_chaining.py

import subprocess
import sys

import pytest


def test_optimized_interpreter_removes_assert_statements():
    completed = subprocess.run(
        [sys.executable, "-O", "-c", "assert False; print('continued')"],
        check=True,
        capture_output=True,
        text=True,
    )

    assert completed.stdout.strip() == "continued"

    # 这是锁定 CPython 的真实优化模式边界：assert 适合内部不变量，不能承担输入校验。


def test_annotations_do_not_check_runtime_value_contracts():
    def positive(value: int) -> int:
        if value <= 0:
            raise ValueError("must be positive")
        return value

    assert positive(2) == 2
    with pytest.raises(TypeError):
        positive("2")
    with pytest.raises(ValueError, match="positive"):
        positive(-2)

    # 字符串失败来自函数体的比较，不是注解检查；正值约束同样必须由运行时代码实现。
